source/tutorials/batch-inference/image-batch-inference.rst

```diff
@@ -0,0 +1,169 @@
+        inputs:
+        - name: model
+          default: s3://valohai-public-files/tutorials/batch-inference/image-batch-inference/model.h5
+        - name: images
+          default: s3://valohai-public-files/tutorials/batch-inference/image-batch-inference/images.zip
+
+What we are doing here is defining a single step for our machine learning pipeline, which is the Batch Inference step. We will run on top of the official ``tensorflow/tensorflow:2.4.1`` Docker image, first install the ``pillow`` and ``valohai-utils`` Python libraries and then run our batch inference code.
+
+Let's test that everything is set up correctly by running on Valohai:
+
+.. code-block:: bash
+
+    vh exec run --adhoc "Batch Inference"
+
+If everything went as planned, we should see our Valohai execution end after finding out that ``batch_inference.py`` is missing:
+
+.. image:: batch-inference-tutorial-1.png
+   :alt: Error, but success!
+
+Unpacking the Images
+--------------------
+
+...actually Valohai unpacks the images automatically, so we can just fetch their location with ``vh.inputs('images').paths()``!
+
+Loading and Using Our Model
+---------------------------
+
+Okey, let's get started by creating and opening up ``batch_inference.py`` in your favorite editor!
+
+Add these imports to the beginning of the file:
+
+.. code-block:: python
+
+    import json
+    import os
+
+    import numpy as np
+    from PIL import Image
+    import tensorflow as tf
+    import valohai as vh
+
+We are using the ``valohai-utils`` library to make integration with Valohai straightforward.
+
+Begin by loading our model:
+
+.. code-block:: python
+
+    model = tf.keras.models.load_model(vh.inputs('model').path())
+
+Easy, huh? Let's define a function to load up an image:
+
+.. code-block:: python
+
+    def load_image(image_path):
+        _, image_name = os.path.split(image_path)
+        image = Image.open(image_path)
+        image.load()
+        image = image.resize((28, 28)).convert('L')
+        image_data = np.array(image).reshape(1, 28, 28)
+        image_data = image_data / 255.0
+        return (image_name, image_data)
+
+Then define a function to run inference on an image:
+
+.. code-block:: python
+
+    def run_inference(image):
+        image_name, image_data = image
+        prediction = np.argmax(model.predict(image_data))
+        with vh.logger() as logger:
+            logger.log('image', image_name)
+            logger.log('inferred_digit', prediction)
+        return {
+            'image': image_name,
+            'inferred_digit': str(prediction),
+        }
+
+Finally, let's run these functions for all the images. While we already log all the individual inference results in the inference function, let's also save the complete results in a single JSON file.
+
+.. code-block:: python
+
+    image_paths = vh.inputs('images').paths()
+
+    images = [load_image(p) for p in image_paths]
+    results = [run_inference(i) for i in images]
```

Comment:
This might not be a good idea for a large dataset, since you're loading all images into memory first, then processing them.

```suggestion
    results = []
    for path in vh.inputs('images').paths():
        results.append(run_inference(load_image(path)))
```

Reply:
We clearly need lazy evaluation! 😄

Reply:
Well, sure:

```
results = [run_inference(i) for i in (load_image(p) for p in vh.inputs('images').paths())]
```

That makes the load-image bit a genexpr, which is lazy. It's not readable, though...

Reply:
😄 yeah, I think your original suggestion is better.